### PR TITLE
feat: Handle restart issue by assuming cameras are in bad state

### PR DESCRIPTION
**Pull request by DrCodeWizzard**
**Please ensure the above *guidelines for contributing* are met.**

Fixes Issue: #

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: realsense_ros_camera/src/realsense_node_factory.cpp [hunk 1]
```diff
@@ -79,6 +79,49 @@ void RealSenseNodeFactory::onInit()
             exit(1);
         }
 
+        // we found the device. Restart the device and wait for it to come up again
```
Comment (mrousseau65) [inline realsense_ros_camera/src/realsense_node_factory.cpp]: Is there anything we can check to see if a device is in a bad state before blindly doing a hardware reset? With the RS200, there were failures that could be detected that indicated it needed to be reset. Avoiding a mandatory 5 second delay at startup if it's not actually needed would be nice.

File: realsense_ros_camera/src/realsense_node_factory.cpp [hunk 2]
```diff
@@ -79,6 +79,49 @@ void RealSenseNodeFactory::onInit()
...
+            ROS_ERROR("No RealSense devices were found! Terminating RealSense Node...");
+            ros::shutdown();
+            exit(1);
+        }
+        found = false;
+        for (auto&& dev : list)
+        {
+            auto sn = dev.get_info(RS2_CAMERA_INFO_SERIAL_NUMBER);
+            auto port_id = parseUsbPortId(dev.get_info(RS2_CAMERA_INFO_PHYSICAL_PORT));
+            if (usb_port_id.empty())
```
Comment (mrousseau65) [inline realsense_ros_camera/src/realsense_node_factory.cpp]: This loop is not guaranteed to select the same device the previous one did (the one that was just reset). So, the returned device may not, in fact, have been reset properly.

File: realsense_ros_camera/src/realsense_node_factory.cpp [hunk 3]
```diff
@@ -79,6 +79,49 @@ void RealSenseNodeFactory::onInit()
...
+            else if (port_id == usb_port_id)
+            {
+                _device = dev;
+                found = true;
+                ROS_INFO_STREAM("Device connected with USB port: " << port_id << " (serial number: " << sn << ") was found.");
+                break;
+            }
+        }
+
+        if (!found)
```
Comment (mrousseau65) [inline realsense_ros_camera/src/realsense_node_factory.cpp:112]: This second loop should guarantee that the device it reset is the one it found and is being returned. This second loop should probably store the SN of the device found first and only select on that condition to ensure this.

**Review by mrousseau65 — CHANGES_REQUESTED** [review]
If more than 1 device is enumerated, you need to make sure you're returning the correct one after a reset.

File: realsense_ros_camera/src/realsense_node_factory.cpp (hunk 1 above)
Comment (DrCodeWizzard) [reply realsense_ros_camera/src/realsense_node_factory.cpp]: Not quite sure on that one. The failures can occur when trying to setup the profile, stream, or trying to retrieve the frame. Based on information I found on github once the device was reset it cannot be used as the pointer is invalid. Thus we would have to restart the nodelet to build the whole software stack from scratch for the camera. This is a rough workaround to give is breathing room to develop something more robust and better.

File: realsense_ros_camera/src/realsense_node_factory.cpp [hunk 4]
```diff
@@ -98,20 +101,13 @@ void RealSenseNodeFactory::onInit()
         {
             auto sn = dev.get_info(RS2_CAMERA_INFO_SERIAL_NUMBER);
             auto port_id = parseUsbPortId(dev.get_info(RS2_CAMERA_INFO_PHYSICAL_PORT));
-            if (usb_port_id.empty())
+            if (sn == serial_no && port_id == usb_port_id)
             {
                 _device = dev;
                 serial_no = sn;
```
Comment (mrousseau65) [inline realsense_ros_camera/src/realsense_node_factory.cpp]: This line is not needed (it is already equal if you made it through the 'if')

**Review by mrousseau65 — APPROVED** [review]
As a bonus, the new code now guarantees 'serial_no' is correct when passed to the device's constructor (wasn't guaranteed in the original code)

**Review by mrousseau65 — APPROVED** [review]

**Pull request merged**